### ISSUE TITLE
Potential bug/confusion when is_multilabel=True for some metrics

### ISSUE THREAD

**Issue by ozancaglayan** [pytorch/ignite#2444]
Hello,

in the documentation for `Precision`, `Recall` and `ClassificationReport`, if the task is multilabel, following examples are given:

```
        Multilabel case, the shapes must be (batch_size, num_categories, ...)

        .. testcode:: 2

            metric = ClassificationReport(output_dict=True, is_multilabel=True)
            metric.attach(default_evaluator, "cr")
            y_true = torch.Tensor([
                [0, 0, 1],
                [0, 0, 0],
                [0, 0, 0],
                [1, 0, 0],
                [0, 1, 1],
            ]).unsqueeze(0)
            y_pred = torch.Tensor([
                [1, 1, 0],
                [1, 0, 1],
                [1, 0, 0],
                [1, 0, 1],
                [1, 1, 0],
            ]).unsqueeze(0)
```

In all metric docs, a first-dim is added by `unsqueeze(0)` to both `y_true` and `y_pred`, but it is also told that the shapes must be `(batch_size, num_categories, ...)`. From my understanding of the latter, both `y_true` and `y_pred` should have `n_samples X num_categories` shape. In this example, this amounts to `5 x 3` e.g. 5 examples/samples and three output labels, each binary.

- Why do we add the dummy singleton dim in the first place? If we don't add it, per-label computation is not correct as it takes the sample dimension as the label dimension i.e. if 256 examples are given, 256 metrics are computed for each of them.

Thanks!

**Comment by vfdev-5** [comment]
@ozancaglayan thanks for the question! 
@sdesrozis can you take a look ?

**Comment by vfdev-5** [comment]
@ozancaglayan I agree that docs look a bit weird. The correct usage should look like in the tests where we match results vs sklearn: https://github.com/pytorch/ignite/blob/master/tests/ignite/metrics/test_classification_report.py#L77-L141

Sorry about that.

**Comment by ozancaglayan** [comment]
even from the tests, I can't understand how this is supposed to work? I get it to work with sklearn as in the tests but:

```
y_true = torch.randint(0, 2, size=(offset * idist.get_world_size(), n_classes, 6, 8)).to(device)
y_preds = torch.randint(0, 2, size=(offset * idist.get_world_size(), n_classes, 6, 8)).to(device)

cr.update([y_preds, y_true])
cr.compute()
```

this does not work either. Is this only supposed to work if you slice the batch dim iteratively but not give that at once? But that shouldn't matter, dunno.

**Comment by vfdev-5** [comment]
@ozancaglayan I checked from my side and agree that `ClassificationReport` metric does not work as I would expect.
Here is how you can compute sample avg precision, recall, f1 meanwhile we are fixing this issue:
```python
from sklearn.metrics import classification_report

import numpy as np

# Multi-label preds/GT
# 4 labels and 8 samples
y_true = [
    [0, 1, 1, 0],
    [1, 0, 1, 0],
    [1, 1, 1, 0],    
    [0, 1, 1, 1],
    [1, 0, 1, 1],
    [1, 1, 1, 1],
    [0, 0, 0, 0],
    [0, 1, 1, 0],    
]
y_preds = [
    [0, 1, 0, 0],
    [1, 0, 1, 0],
    [1, 1, 0, 0],    
    [0, 1, 0, 1],
    [1, 0, 0, 1],
    [1, 1, 0, 1],
    [1, 0, 1, 0],
    [0, 1, 1, 0],
]


true_result = classification_report(y_true, y_preds, output_dict=True)
print(true_result['samples avg'])

import torch
from ignite.metrics import Precision, Recall, Fbeta
from ignite.engine import Engine


t_y_true = torch.tensor(y_true)
t_y_preds = torch.tensor(y_preds)

evaluator = Engine(lambda e, b: b)

pr = Precision(is_multilabel=True, average=False)
re = Recall(is_multilabel=True, average=False)
f1 = Fbeta(beta=1, precision=pr, recall=re)
avg_pr = pr.mean()
avg_re = re.mean()

avg_pr.attach(evaluator, "sample_avg_precision")
avg_re.attach(evaluator, "sample_avg_recall")
f1.attach(evaluator, "sample_avg_f1")

bs = 4
evaluator.run([
    [t_y_preds[:bs], t_y_true[:bs]],
    [t_y_preds[bs:], t_y_true[bs:]]    
])

print(evaluator.state.metrics)
```
Please, let me know if it works for you. Thanks

**Comment by louis-she** [comment]
assign to me pls

**Comment by sdesrozis** [comment]
Sorry, I completely missed that question!

**Comment by louis-she** [comment]
Sorry i'm a little busy these days, i have not started this one yet. @nishantb06 you can take this one. I will focus on #2423

**Comment by sadra-barikbin** [comment]
@vfdev-5  I'll do it, if you will

**Comment by vfdev-5** [comment]
Closing this issue as done by https://github.com/pytorch/ignite/pull/2573

@ozancaglayan you can check the fix on our nightly release : `pip install --upgrade --pre pytorch-ignite` , if you'd like. Thanks!
